system/jlib/jutil.cpp

```diff
@@ -3056,19 +3175,481 @@ extern jlib_decl void setSecretMount(const char * path)
+        else
+            role.set("hpcc-vault-access");
+        if (vault->hasProp("@client-secret"))
+        {
+            useKubernetesAuth = false;
+            //for now only support direct access token.  we can support other combinations for example login token, ldap login, etc later.
+            Owned<IPropertyTree> clientSecret = getLocalSecret(vault->queryProp("@client-secret"));
+            token.set(clientSecret->queryProp("token"));
+        }
+    }
+    CVaultKind getVaultKind() const { return kind; }
+    void kubernetesLogin()
+    {
+        CriticalBlock block(vaultCS);
+        if (token.length())
+            return;
+        StringBuffer login_token;
+        login_token.loadFile("/var/run/secrets/kubernetes.io/serviceaccount/token");
+        if (login_token.length())
+        {
+            std::string json;
+            json.append("{\"jwt\": \"").append(login_token.str()).append("\", \"role\": \"").append(role.str()).append("\"}");
+            httplib::Client cli(schemeHostPort.str());
+            if (username.length() && password.length())
+                cli.set_basic_auth(username, password);
+            if (httplib::Result res = cli.Post("/v1/auth/kubernetes/login", json, "application/json"))
+            {
+                if (res->status == 200)
+                {
+                    const char *response = res->body.c_str();
+                    if (!isEmptyString(response))
+                    {
+                        Owned<IPropertyTree> respTree = createPTreeFromJSONString(response);
+                        if (respTree)
+                            token.set(respTree->queryProp("auth/client_token"));
+                    }
+                }
+                else
+                {
+                    Owned<IException> e = MakeStringException(0, "Vault kube auth error [%d](%d) - vault: %s - response: %s", res->status, res.error(), name.str(), res->body.c_str());
+                    OWARNLOG(e);
+                    throw e.getClear();
+                }
+            }
+        }
+    }
+    bool getCachedSecret(CVaultKind &rkind, StringBuffer &content, const char *secret, const char *version)
     {
-        CriticalBlock block(secretCS);
-        if (secretDirectory.isEmpty())
-            setSecretMount(nullptr);
+        unsigned timeoutThreshold = msTick() - getSecretTimeout();
+        CriticalBlock block(vaultCS);
+        IPropertyTree *tree = cache->queryPropTree(secret);
+        if (tree)
+        {
+            VStringBuffer vername("v.%s", isEmptyString(version) ? "latest" : version);
+            IPropertyTree *envelope = tree->queryPropTree(vername);
+            if (!envelope)
+                return false;
+            unsigned created = (unsigned) envelope->getPropInt("@created");
+            if (created && (created < timeoutThreshold))
+            {
+                tree->removeTree(envelope);
+                puts("\nremoved from vault cache\n");
```

Comment:
similar comments about tracing with puts

Reply:
Forgot to scrub these personal debug statements out.